message_converter/include/converter_node.hpp

```diff
@@ -0,0 +1,106 @@
+#pragma once
+#include "ros/ros.h"
+#include <ros/console.h>
+#include <boost/array.hpp>
+#include <geometry_msgs/PointStamped.h>
+#include <gnss_msgs/BaselinePosition.h>
+#include <gnss_msgs/BaselineVelocity.h>
+#include <gnss_msgs/RtkAvailable.h>
+#include <geometry_msgs/PointStamped.h>
+#include <piksi_rtk_msgs/BaselineNed.h>
+#include <piksi_rtk_msgs/VelNedCov.h>
+#include <libsbp_ros_msgs/MsgBaselineNed.h>
+#include <libsbp_ros_msgs/MsgVelNedCov.h>
+#include <libsbp_ros_msgs/MsgPosLlh.h>
+#include <eigen3/Eigen/Dense>
+#include <math.h>
+#include <std_msgs/Bool.h>
+
+#include <diagnostic_msgs/DiagnosticArray.h>
+#include <diagnostic_updater/diagnostic_updater.h>
+#include <diagnostic_updater/publisher.h>
+
+class SwiftNavRover{
+public:
+    SwiftNavRover(ros::NodeHandle* node_handle, double la, double lo, double al);
+
+    // WGS 84 derived geometric constants
+    const double kSemimajorAxis = 6378137.0;
+    const double kSemiminorAxis = 6356752.3142;
+    const double kFirstEccentricitySquared = 6.69437999014 * 0.001;
+    const double kSecondEccentricitySquared = 6.73949674228 * 0.001;
+    const double kFlattening = 1/298.257223563;
+
+    typedef Eigen::Matrix<double,3,3> Matrix3x3d;
+    Matrix3x3d ecef_to_ned_matrix;
+
+    ros::NodeHandle nh;
+    
+    void init();
+
+    // Publisher functions
+    void publishBaselinePosition(ros::Time t, double n, double e, double d, int tow, boost::array<double, 9> covariance, int n_sats, int fixed_mode);
+    void baselinePositionCallback(const libsbp_ros_msgs::MsgBaselineNed::ConstPtr & msg);
+    void publishBaselineVelocity(ros::Time t, int tow,double n,double e,double d,boost::array<double, 9> covariance,int n_sats,int vel_mode,int ins_mode);
+    void publishRTKAvailable(ros::Time t);
+    void publishStatus(const ros::TimerEvent& event);
+
+    // Callback functions
+    void baselineVelocityCallback(const libsbp_ros_msgs::MsgVelNedCov::ConstPtr & msg);
+    void posLLHCallback(const libsbp_ros_msgs::MsgPosLlh::ConstPtr & msg);
+
+    // Co-ordinate conversions
+    void geodetic2ned(double latitude, double longitude, double altitude, double* north, double* east, double* down);
```

Comment:
most if not all pointers in this class interface can in fact be replaced with references (&) which are a bit safer in that the location they point to cannot be changed.